### ISSUE TITLE
Inconsistent results of sendReportedProperties method multiple call

### ISSUE THREAD

**Issue by BratnerMichal** [Azure/azure-iot-sdk-java#207]
- **OS and version used:** Linux Kernel 3.2.0 on embedded device

- **Java runtime used:** 1.8.0_144

- **SDK version used:** iot-device-client 1.8.0

# Description of the issue:
Hello, I want to use IotHub Device Twins to store properties from embedded device. I'm testing update properties by `sendReportedProperties()` method. However I noticed that when I call this method multiple times in short period, not every change is always propagated to IotHub. 

When I run the same application with the same properties changes repeatedly, final properties state on IotHub is almost always different. This inconsistent behavior looks like sometimes some `sendReportedProperties()` method calls are not processed on IotHub but every time I get `OK_EMPTY` response for every request.

# Code sample exhibiting the issue:

Internal `IotHubProperty` object
```
public class IotHubProperty {
   public String key;
   public Object value;
	
   public IotHubProperty() {
   }
	
   public IotHubProperty(String key, Object value) {
      this.key = key;
      this.value = value;
   }

   public String getKey() {
      return this.key;
   }
	
   public Object getValue() {
      return this.value;
   }
}
```

Method that tries update of reported properties
```
private void sendTestProperties() throws IOException {
   DeviceClient client = new DeviceClient(connectionString, IotHubClientProtocol.MQTT);
   IotDeviceManager deviceManager = new IotDeviceManager(client);
 
   deviceManager.clientConnect();

   IotHubProperty prop1 = new IotHubProperty("key1", "value1");
   IotHubProperty prop2 = new IotHubProperty("key2", "value2");

   deviceManager.setProperty(prop1);
   deviceManager.setProperty(prop2);

   prop1.value = "new value1";
   prop2.value = "new value2";

   deviceManager.setProperty(prop2);
   deviceManager.setProperty(prop1);
}
```

Class which manages sending properties to IotHub
```
public class IotDeviceManager {
   List<IotHubProperty> reported = new ArrayList<IotHubProperty>();  // List of properties in internal objects
   DeviceClient client;
   DesiredPropertiesListener telemetryConfig = new DesiredPropertiesListener();

   public IotDeviceManager(DeviceClient client) {
      this.client = client;
   }

   public void clientConnect() throws IOException {
      client.open();
      client.startDeviceTwin(new DeviceTwinStatusCallBack(), null, telemetryConfig, null);
      client.subscribeToDesiredProperties(telemetryConfig.getDesiredProp());
    }

   public void setProperty(IotHubProperty property) {
      for(IotHubProperty prop: reported) {
         if(prop.getKey().equals(property.getKey())) {
            // Exchange for property with new value
    	    reported.remove(prop);
            reported.add(property);
            sendReportedProperties();
            return;
         }
      }
      reported.add(property);
      sendReportedProperties();
   }

   private void sendReportedProperties() {
      for(IotHubProperty property: reported) {
         telemetryConfig.setReportedProp(new Property(property.getKey(), property.getValue()));
      }
      try {
         client.sendReportedProperties(telemetryConfig.getReportedProp());
      } catch (IOException e) {
         // Handle the exception
      }
   }
   
   private class DesiredPropertiesListener extends Device 
   {
      @Override
      public void PropertyCall(String propertyKey, Object propertyValue, Object context) 
      {
         // Reading desired properties, it works fine
      }
   }

   protected static class DeviceTwinStatusCallBack implements IotHubEventCallback {
      public void execute(IotHubStatusCode status, Object context) {
         System.out.println("IoT Hub responded to device twin operation with status " + status.name());
      }
   }
}
```

# Final properties state on IotHub:
After this part of code, I would expect on IotHub properties in this state:
```
"key1":"new value1",
"key2":"new value2"
```

But sometimes there is
```
"key1":"value1",
"key2":"value2"
```

or 
```
"key1":"value1",
"key2":"new value2"
```

Are there some request count or properties update count limits per time which may cause that some of later updates are not sometimes processed? If so, I would expect negative response from IotHub after exceeding these limits.

I also tried sort properties by key before send and add 2 seconds timeout after `client.sendReportedProperties()` method but nothing helps.

We use S1 Azure IotHub unit.

**Comment by jasmineymlo** [comment]
@BratnerMichal Thanks for filing the issue.  Would you please tried with the latest release?  A recent change 00ca4d35c1f1ad4718ac8ae56eb681d1fa89c33e fixed reporting the actual response of message received and process rather than only on delivery complete.  Please let us know the result.

Moreover, I try to repo with modification of the DeviceTwinSample, calling sendReportedProperties nultiple times with different property values but wasn't able to get a repo...

**Comment by BratnerMichal** [comment]
@jasmineymlo What I know, the latest release is 1.8.0 and I reported the issue with this version.

However, I found solution for my problem. When I call `clean()` method for `Device` object after every `sendReportedProperties()` call, results are then consistent and correct. 

It does not explain inconsistent behavior but solves my problem.

**Comment by jasmineymlo** [comment]
@BratnerMichal.  Thanks for the info.  Since calling clean() is the workaround for the issue and unblocks you for now.  We will need to further investigate into the inconsistent behavior.

**Comment by BratnerMichal** [comment]
@jasmineymlo Yes, it works for me now. However, I noticed one thing which maybe could help you with investigation. When I don't call `clean()` method, then after each `setReportedProp()` call are new properties stacking in Reported properties set regardless of whether there is already other property with the same key.

So in my case without `clean()` method, at the end the Reported properties set contained two properties with key `key1` and two properties with key `key2`. That means when I add new property with the same key as any already existing property, it does not replace the old one but is added as new one.

I don't know whether this behavior has some reason but I would expect that replacing properties with the same key is cleaner way than duplicate them.

**Comment by jasmineymlo** [comment]
Thanks for the info.  We will take a look at it when we get a chance.

**Comment by timtay-microsoft** [comment]
Right now, the SDK does allow you to set a property twice with different values, but only one of them will actually be reported when you call sendReportedProperties. It's possible that you are not ensuring that the set of properties you are sending has no duplicates. I'm adding a check in the SDK to help catch cases like this in the future.

**Comment by timtay-microsoft** [comment]
The fix for this bug has been released as of Device Client 1.13.0. If you still see this behavior after upgrading, feel free to re-open this issue
